### PR TITLE
Add support for reloadSources, IsolateReload and PausePostRequest

### PR DESCRIPTION
**Pull request by BlackHC**
This was tested with Dart 1.21.1.
Remaining TODOs:
* semantic version update?
* Dart SDK constraints?
* tests for PausePostRequest and IsolateReload

PTAL! I don't know much about the remaining TODOs. Some advice would be awesome! :)

Thanks,
 Andreas

### CODE REVIEW

File: lib/src/isolate.dart [hunk 1]
```diff
@@ -69,6 +70,11 @@ class VMIsolateRef {
   Stream<VMIsolateRef> get onUpdate => _onUpdate;
   Stream<VMIsolateRef> _onUpdate;
 
+  /// A broadcast stream that emits a new reference to this isolate every time
+  /// it has been reloaded.
```
Comment (nex3) [inline lib/src/isolate.dart]: Explain what "reload" means in this context.

Why is this returning a new isolate ref? Does a reload change the metadata at all?

Also add a note indicating what protocol version this was added in.

File: lib/src/isolate.dart [hunk 2]
```diff
@@ -318,6 +329,35 @@ class VMIsolateRef {
     }
   }
 
+  Future<VMReloadReport> reloadSources(
```
Comment (nex3) [inline lib/src/isolate.dart]: Document this. In general, we try to make our documentation at least as good as the protocol documentation, and better where possible.

Also add a note indicating what protocol version this was added in.

File: lib/src/isolate.dart [hunk 3]
```diff
@@ -318,6 +329,35 @@ class VMIsolateRef {
...
+      {bool force, bool pause, rootLibUri, packagesUri}) async {
```
Comment (nex3) [inline lib/src/isolate.dart]: Name these "uri" parameters "url" instead. "Uri" is basically a deprecated term at this point.

File: lib/src/isolate.dart [hunk 4]
```diff
@@ -318,6 +329,35 @@ class VMIsolateRef {
...
+    if (rootLibUri != null && rootLibUri is! String && rootLibUri is! Uri) {
+      throw new ArgumentError(
```
Comment (nex3) [inline lib/src/isolate.dart]: Use `new ArgumentError.value()`.

File: lib/src/pause_event.dart [hunk 5]
```diff
@@ -103,6 +103,13 @@ class VMPauseExceptionEvent extends VMPauseEvent {
   String toString() => "pause on exception";
 }
 
+/// An event indicating that an isolate was paused after a service request.
+class VMPausePostRequestEvent extends VMPauseEvent {
```
Comment (nex3) [inline lib/src/pause_event.dart]: It looks like this isn't being constructed anywhere...

File: lib/src/reload_report.dart [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
```
Comment (nex3) [inline lib/src/reload_report.dart]: 2017

File: lib/src/reload_report.dart [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+VMReloadReport newVMReloadReport(Map json) {
+  if (json == null) return null;
+  assert(json["type"] == "ReloadReport");
+  return new VMReloadReport._(json);
+}
+
+/// The version of the VM service protocol (*not* the Dart VM itself).
+class VMReloadReport {
```
Comment (nex3) [inline lib/src/reload_report.dart]: This needs to be exported from `lib/vm_service_client.dart`.

File: lib/src/reload_report.dart [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+VMReloadReport newVMReloadReport(Map json) {
+  if (json == null) return null;
+  assert(json["type"] == "ReloadReport");
+  return new VMReloadReport._(json);
+}
+
+/// The version of the VM service protocol (*not* the Dart VM itself).
```
Comment (nex3) [inline lib/src/reload_report.dart]: This comment isn't accurate.

File: test/isolate_test.dart [hunk 9]
```diff
@@ -67,6 +67,14 @@ void main() {
       await main.setName('fblthp');
     });
 
+    test("reloadSources can succeed", () async {
+      var client = await runAndConnect();
+      var isolate = await (await client.getVM()).isolates.first.loadRunnable();
+
+      final report = await isolate.reloadSources(force: true);
+      expect(report.status, true);
+    }, tags: ["reload"]);
```
Comment (nex3) [inline test/isolate_test.dart]: We don't really use tags in this package, so tagging only this test is strange. In general, I don't use tags unless I have a specific reason to—either wanting to select a subset of tests that appear across many different files, or if I want to add special metadata for that tag.

**Review by nex3 — CHANGES_REQUESTED** [review]
Thanks for the patch!

> semantic version update?

Yes, definitely. You're adding new features, so the version should be bumped to 0.2.4 and a changelog entry should be added.

> Dart SDK constraints?

We keep the lower SDK constraint tacked to the version that has the earliest VM service version we support, so that shouldn't change.

> tests for PausePostRequest and IsolateReload

These definitely need to exist. Look at the existing tests for inspiration.

**Comment by BlackHC** [comment]
Reviewed 4 of 4 files at r1, 6 of 7 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[lib/src/isolate.dart, line 74 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntD-r1-74:-KeFlNCdK8rOr0vjuZvM:b-1l47tj) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/isolate.dart#L74)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

Explain what "reload" means in this context.

Why is this returning a new isolate ref? Does a reload change the metadata at all?

Also add a note indicating what protocol version this was added in.
</blockquote></details>

All of this was added at protocol version 3.6 (unreleased atm?), see bottom of https://github.com/dart-lang/sdk/blob/ebe7d148f3964b17f1575a731e7d32e96855258d/runtime/vm/service/service.md 
I'm not sure how to treat this.

And good question. I don't think so. I've replaced it with Null now. This is purely if you want to listen to reload events.

---

*[lib/src/isolate.dart, line 332 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntD-r1-332:-KeFoU5ewOHIRmjg0J3A:b-896fix) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/isolate.dart#L332)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

Document this. In general, we try to make our documentation at least as good as the protocol documentation, and better where possible.

Also add a note indicating what protocol version this was added in.
</blockquote></details>

Done.

---

*[lib/src/isolate.dart, line 333 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntD-r1-333:-KeFprFDz0vknYcbCnb2:b-896fix) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/isolate.dart#L333)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

Name these "uri" parameters "url" instead. "Uri" is basically a deprecated term at this point.
</blockquote></details>

Done.

---

*[lib/src/isolate.dart, line 335 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntD-r1-335:-KeFqDa5SarSK8FeG43c:b-wtwdix) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/isolate.dart#L335)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

Use `new ArgumentError.value()`.
</blockquote></details>

Done. Pretty cool!

---

*[lib/src/reload_report.dart, line 1 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntF-r1-1:-KeFqnkHBUfNTKsSBzAt:b-896fix) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/reload_report.dart#L1)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

2017
</blockquote></details>

Done.

---

*[lib/src/reload_report.dart, line 11 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntF-r1-11:-KeFqoH2d-p31M0v-jbC:b-896fix) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/reload_report.dart#L11)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

This comment isn't accurate.
</blockquote></details>

Done.

---

*[lib/src/reload_report.dart, line 12 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntF-r1-12:-KeFr1kbgdWmtqAdoLKe:b-896fix) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/reload_report.dart#L12)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

This needs to be exported from `lib/vm_service_client.dart`.
</blockquote></details>

Done.

---

*[test/isolate_test.dart, line 76 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntG-r1-76:-KeFr3JX3SiStd-F6n8d:bncypd6) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/test/isolate_test.dart#L76)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

We don't really use tags in this package, so tagging only this test is strange. In general, I don't use tags unless I have a specific reason to—either wanting to select a subset of tests that appear across many different files, or if I want to add special metadata for that tag.
</blockquote></details>

Done. Sorry.

---

*[lib/src/pause_event.dart, line 107 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntE-r1-107:-KeFqROX6XjTMej7tkXT:b-896fix) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/pause_event.dart#L107)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

It looks like this isn't being constructed anywhere...
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14)*

**Comment by BlackHC** [comment]
Thanks for the review PTAL. The test to check that onReload works actually surfaced an issue. It only triggers for failed reloads, not for successful ones. I've filed an issue at https://github.com/dart-lang/sdk/issues/28960.

How shall we proceed?

Thanks,
 Andreas

**Comment by nex3** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[lib/src/isolate.dart, line 74 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntD-r1-74:-Ke_S3nw0vFuECqLj9eq:b-aoxo3l) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/isolate.dart#L74)):*
> I'm not sure how to treat this.

Your change looks good.

> I've replaced it with Null now.

This package doesn't use `<Null>`. I generally like following the spec-mode convention of leaving off the type annotation, at least until the strong mode semantics of `<Null>` and `<void>` are sorted out for good.

---

*[lib/src/isolate.dart, line 332 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntD-r1-332:-Ke_T25mqEGUP0tG37VZ:bdet0uj) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/isolate.dart#L332)):*
<details><summary><i>Previously, BlackHC (Andreas Kirsch) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It's still not really clear what it means to "reload sources". Which sources are affected? What changes? Does it affect already-instantiated objects or only new ones?

---

*[test/isolate_test.dart, line 77 at r3](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-Ke_Ubnm6taKd9Ko6x6e:-Ke_Ubnm6taKd9Ko6x6f:b1evstq) ([raw file](https://github.com/dart-lang/vm_service_client/blob/1f4cf23d5dfb8fb2c0cc482191c34bd791315246/test/isolate_test.dart#L77)):*
> ```Dart
>       // right now, issue tracked at:
>       // https://github.com/dart-lang/sdk/issues/28960
>       final report = await isolate.reloadSources(force: true);
> ```

Style nit: use `var` for local variables.

---

*[test/isolate_test.dart, line 79 at r3](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-Ke_UECJUCEvtoeG7GVf:-Ke_UECJUCEvtoeG7GVg:bgmnz8f) ([raw file](https://github.com/dart-lang/vm_service_client/blob/1f4cf23d5dfb8fb2c0cc482191c34bd791315246/test/isolate_test.dart#L79)):*
> ```Dart
>       final report = await isolate.reloadSources(force: true);
>       expect(report.status, true);
>     });
> ```

This should also check that it actually reloads the sources—construct a newly-added class or something.

Also, make sure to test each parameter.

---

*[test/isolate_test.dart, line 87 at r3](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-Ke_UgcA6hhuE4QXCLqe:-Ke_UgcA6hhuE4QXCLqf:bi2ctod) ([raw file](https://github.com/dart-lang/vm_service_client/blob/1f4cf23d5dfb8fb2c0cc482191c34bd791315246/test/isolate_test.dart#L87)):*
> ```Dart
>       // TODO(blackhc): IsolateReload provides more information which we do
>       // not surface currently. We should verify the error type, too.
>       expect(isolate.onReload.first, completion(isNull));
> ```

It might be nice to verify that this only fires once. Using test's new stream matchers I think you can write `expect(isolate.onReload, emitsInOrder([isNull, neverEmits(anything)]))`.

---

*[test/isolate_test.dart, line 110 at r3](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-Ke_U6kMCSyEmPPMXBFt:-Ke_U6kNo4g1mWdHk3W9:b-25u0oj) ([raw file](https://github.com/dart-lang/vm_service_client/blob/1f4cf23d5dfb8fb2c0cc482191c34bd791315246/test/isolate_test.dart#L110)):*
> ```Dart
>       var queue = new StreamQueue(main.onPauseOrResume);
>       expect(
>           queue.next, completion(new isInstanceOf<VMPauseInterruptedEvent>()));
> ```

Please remove unrelated formatting changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14)*

**Comment by BlackHC** [comment]
Hey! I've run into some other things that I need for the REPL that are undocumented in the spec, so I don't feel like I can add them to the library/this CL.  I can't prioritize the review over just using it and finishing the next iteration. Could you take over the remainder of the CL maybe? I won't be able to look at this in the near future :( I'm just going to use a separate poc git branch for now. Sorry!! :(

Thanks,
Andreas

---

Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[lib/src/isolate.dart, line 332 at r1](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-KeFivOw7fbG1KW9LntD-r1-332:-KfLL0nTv6HMu9puS5b8:b-bmqiyk) ([raw file](https://github.com/dart-lang/vm_service_client/blob/32323438576b67791470d5e5a3444725ce1daed2/lib/src/isolate.dart#L332)):*
<details><summary><i>Previously, nex3 (Natalie Weizenbaum) wrote…</i></summary><blockquote>

It's still not really clear what it means to "reload sources". Which sources are affected? What changes? Does it affect already-instantiated objects or only new ones?
</blockquote></details>

To be honest, I don't really know myself. This is what's explained in the VM service spec :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dart-lang/vm_service_client/14#-:-KfLLE1QNaebZjM8nL3_:bcdyhzg)*

**Comment by nex3** [comment]
I don't currently have time to put into this either, but I'll leave it open for now until I do or until someone else wants to adopt it.

**Pull request closed without merging**